litestar/app.py

```diff
@@ -482,6 +491,57 @@ async def __call__(
         scope["state"] = {}
         await self.asgi_handler(scope, receive, self._wrap_send(send=send, scope=scope))  # type: ignore[arg-type]
 
+    async def _call_lifespan_handler(self, handler: LifeSpanHandler) -> None:
+        """Determine whether the lifecycle handler expects an argument, and if so pass the `app.state` to it. If the
+        handler is an async function, await the return.
+
+        Args:
+            handler: sync or async callable that may or may not have an argument.
+        """
+        async_callable = AsyncCallable(handler)  # type: ignore[arg-type]
+
+        if async_callable.num_expected_args > 0:
+            await async_callable(self.state)  # type: ignore[arg-type]
+        else:
+            await async_callable()  # pyright: ignore
+
+    @asynccontextmanager
+    async def lifespan(self) -> AsyncGenerator[None, None]:
+        """Context manager handling the ASGI lifespan.
+
+        It will be entered when the ``lifespan`` message has been received from the
+        server, and exit after the ``asgi.shutdown`` message. During this period, it is
+        responsible for calling the ``before_startup``, ``after_startup``,
+        `on_startup``, ``before_shutdown``, ``on_shutdown`` and ``after_shutdown``
+        hooks, as well as custom lifespan managers.
+        """
+        async with AsyncExitStack() as exit_stack:
+            exit_callbacks: list[Callable[[], Awaitable[Any]]] = [
+                *[partial(hook, self) for hook in self.after_shutdown[::-1]],
+                *[partial(self._call_lifespan_handler, handler) for handler in self.on_shutdown[::-1]],
+                self.event_emitter.on_shutdown,
+                *[partial(hook, self) for hook in self.before_shutdown[::-1]],
+            ]
```

Comment:
This is a bit awkward currently, but will be cleaned up when we refactor the hook signatures.